### ISSUE TITLE
Potential Bug: Download extension code from OnPremise server use ONLY Object Names in filenames inside .ZIP

### ISSUE THREAD

**Issue by JavierFuentes** [microsoft/AL#4225]
I have uploaded a few modifications in the form of an extension to my Business Central OnPremise server and when I download them manually, I have noticed that the .ZIP file contains some of the files that I have in my project renamed using _ before and after the file and also they are all on the same root route ...

![image](https://user-images.githubusercontent.com/5801847/48671792-ca9ad300-eb2d-11e8-8662-ec11b26bfacd.png)

I wonder what would happen if I had 2 files in my project located in different subfolders (for example 1 Table and 1 Page) with exactly the same name ...

The logic says that being of different types can have the same name and the same ID but when downloading the source code, in the ZIP surely one of the 2 files would not be ...

Is it a potential Bug?
The internal structure of the code should not respect the subfolders in which each file was contained in the original project?

See also my proposal #4224 to set an official structure and nomenclatures for the AL Extensions projects.

Doing it later will mean a lot of people having to revise their projects, so I think it's better to do it as soon as possible.

**Comment by AskeHolst** [comment]
If you follow the [official best practices](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/compliance/apptest-bestpracticesforalcode) there will be no duplicate file names.

**Comment by JavierFuentes** [comment]
Hi @AskeHolst , I refer to file names inside the ZIP you download from server metadata... __the file names there are build from name of Objects and not from my original project names, so it doesn't really matter how you name your files in your project but it matters how you name your objects in a unique way although they are of different types__...

After posting this issue, I've made a test changing the page extension object "Customer Price Group(s) Ext" to match with table extension object "Customer Price Group Ext" and when I downloaded from my Server, the ZIP archive includes 2 files with the exact equals name _Customer Price Group Ext_ but when I tried to uncompress it, I had an (obvious) File System I/O error .

**Comment by AskeHolst** [comment]
@JavierFuentes you are right, I did not realize the difference, I mistakenly thought the server kept the .app file for later download when showmycode was set. 
I can only agree with your suggestion to do something about this. I have an app with a "SomeFunction Setup" table and a "SomeFunction Setup" page which is the way naming has always been done. That gives an "un-unzippable" zip file.
Having the BC server follow the best practices for file naming would be a good start.

**Comment by StanislawStempin** [comment]
Closing the issue since I see it neither as a bug report or a feature suggestion. You should not rely in any way on the internal structure of the package.

**Comment by JavierFuentes** [comment]
Hi @StanislawStempin,
please notice too that in the image above, the name **file0_D365_BUS_PREMIUM.xml** inside the ZIP is not the name of this _PermissionSets_ file in my project... in my project this file is under a _Permissions_ sub-folder but system has added **file0_** prefix and put it on the root folder of the ZIP

**Comment by rvanbekkum** [comment]
This turns out to be an issue for one of our extensions which has files with object definitions with the same object name, but different object types. If you download the source code from the database, then you don't get all the source files. So, for example, the source code for a table will be downloaded, but the source code of the page with the same object name won't.

**Comment by qutreson** [comment]
Hi @JavierFuentes, can you try to reproduce this issue on the latest developer preview? If this reproduces, can you attach a small AL package that reproduces the issue?

**Comment by JavierFuentes** [comment]
Hi @qutreson,
sorry but we are a final customer with a final customer solutions developer license that seems doesn't work with the latest v15.x OnPrem versions...

Is there any Developer Preview License we can use with Spanish OnPrem version of BC?

**Comment by kalberes** [comment]
This is a bug, indeed.

**Comment by ALGitHubBot** [comment]
The fix for this issue has been checked in to the master branch. It will be available in the bcinsider.azurecr.io/bcsandbox-master Docker image starting from platform build number 38269. 
 
 If you don’t have access to these images you need to become part of the Ready2Go program: aka.ms/readytogo 
 
 For more details on code branches and docker images please read: 
 https://blogs.msdn.microsoft.com/nav/2018/05/03/al-developer-previews-multiple-releases-and-github/ 
 https://blogs.msdn.microsoft.com/freddyk/2018/04/16/which-docker-image-is-the-right-for-you/

**Comment by JavierFuentes** [comment]
I just tested it again with 3 files inside a subfolder of muy project:
./Reports/MyReport.al
./Reports/MyReport.docx
./Reports/MyReport.rdlc

I've deployed my extension into a last version v19. docker container.

When I download my sources, the .ZIP file has MyReport.al at root folder, but keeps the other files at ./Reports path

Definitely, the sources download .ZIP file does not fully respect the original folder structure of the project, and therefore, the file loss problem has not been fully resolved.
